### PR TITLE
feat: Sort by any metadata field

### LINKED ISSUE #5: Sort, search and filter
Search and filter based on image names, modified times and image-wise labels.
Sort images based on ...ditto...

### PR DESCRIPTION
**Pull request by SilviaZeta**
# Description

Adds:

- Left drawer (moved image labels there).
- Sort functionality: from the search bar, after selecting a metadata key it is possible to order the images in ascending or descending order for the key selected.

closes #5,  #27

# Dependency changes
No

# Testing
No

# Documentation
N/A

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by ChasNelson1990** [comment]
Can we use:
`import SortIcon from '@material-ui/icons/Sort';`
As the icon for the sort options?

What if I want to filter by one field and sort by another? Could these be separated?

File: src/ui.tsx [hunk 1]
```diff
@@ -96,9 +89,44 @@ export class UserInterface extends Component<Props, State> {
...
+        return -1;
+      }
+      if (a > b) {
+        return 1;
+      }
+      return 0;
+    }
+
+    this.setState((prevState) => {
+      if (key.toLowerCase().includes("date")) {
```
Comment (cooper667) [inline src/ui.tsx:110]: You could detect whether the _value_ is a date as they will be ISO strings? 

Also, maybe pass `sort` into the compare func, and do any type casting there? Will let you pass anything to the same function. 

I think in practice, dates won't be strings here, they will be actual Date objects, so we can have them as as type?


(happy to merge as is tho tbh)

**Review by cooper667 — APPROVED** [review]

**Comment by SilviaZeta** [comment]
> Can we use:
> `import SortIcon from '@material-ui/icons/Sort';`
> As the icon for the sort options?
Ok
 
> What if I want to filter by one field and sort by another? Could these be separated?
You can do that by searching for one field, then selecting a different field and applying the sort. A search action does not undo a sort (and vice versa). Though we cannot combine more than one search (or sort).

File: src/ui.tsx [hunk 2]
```diff
@@ -96,9 +89,44 @@ export class UserInterface extends Component<Props, State> {
...
-    this.setState({
-      imageNames: this.getImageNames(filteredMeta),
+    this.setState({ filteredMeta });
+  };
+
+  handleOnSortSubmit = (key: string, sortOrder: string): void => {
+    // Handle sort by any string or by date.
+    const sort = sortOrder === "asc" ? 1 : -1;
+
+    function compare(a: string | Date, b: string | Date): number {
```
[the remark this answers is not in the document]
Comment (SilviaZeta) [reply src/ui.tsx:99]: Could do.. at the moment all metadata fields are either strings or array of strings, so supporting numbers didn't seem necessary..

File: src/ui.tsx (hunk 1 above)
Comment (SilviaZeta) [reply src/ui.tsx:110]: Makes sense! At the moment the metadata comes from a temporary metadata.json file we import. The way in which we store this data may change, so I'll be tempted to open a new issue and reconsider this when the data types are final?

Comment (SilviaZeta) [reply src/ui.tsx:110]: Ok for passing `sort` in the compare function.

Comment (cooper667) [reply src/ui.tsx:110]: > Makes sense! At the moment the metadata comes from a temporary metadata.json file we import. The way in which we store this data may change, so I'll be tempted to open a new issue and reconsider this when the data types are final?

Yep

**Comment by ChasNelson1990** [comment]
> > What if I want to filter by one field and sort by another? Could these be separated?
> You can do that by searching for one field, then selecting a different field and applying the sort. A search action does not undo a sort (and vice versa). Though we cannot combine more than one search (or sort).

Hmmm... but then there's no way to see what filters you have applied is there?

**Comment by SilviaZeta** [comment]
> Hmmm... but then there's no way to see what filters you have applied is there?
Correct..

**Review by cooper667 — APPROVED** [review]

**Review by ChasNelson1990 — COMMENTED** [review]
Will make a couple of follow up issues but happy for this to be merged

**Review by ChasNelson1990 — APPROVED** [review]

**Pull request merged**